crypto/fipsmodule/ec/p384.c

```diff
@@ -56,18 +31,60 @@ static const p384_felem p384_felem_one = {
 //
 #if !defined(OPENSSL_NO_ASM) && defined(OPENSSL_LINUX) && \
     (defined(OPENSSL_X86_64) || defined(OPENSSL_AARCH64))
-#include "../../../third_party/s2n-bignum/include/s2n-bignum_aws-lc.h"
+
+#  include "../../../third_party/s2n-bignum/include/s2n-bignum_aws-lc.h"
+
+#  define P384_USE_S2N_BIGNUM_FIELD_ARITH 1
+#  define P384_USE_64BIT_LIMBS_FELEM 1
+
+#else
+
+#  if defined(BORINGSSL_HAS_UINT128)
+#    include "../../../third_party/fiat/p384_64.h"
+#    define P384_USE_64BIT_LIMBS_FELEM 1
+#  else
+#    include "../../../third_party/fiat/p384_32.h"
+#  endif
+
+#endif
+
+#if defined(P384_USE_64BIT_LIMBS_FELEM)
+
+#define P384_NLIMBS (6)
+typedef uint64_t p384_limb_t;
+typedef uint64_t p384_felem[P384_NLIMBS];
+static const p384_felem p384_felem_one = {
+    0xffffffff00000001, 0xffffffff, 0x1, 0x0, 0x0, 0x0};
+
+#else  // 64BIT; else 32BIT
+
+#define P384_NLIMBS (12)
+typedef uint32_t p384_limb_t;
+typedef uint32_t p384_felem[P384_NLIMBS];
+static const p384_felem p384_felem_one = {
+    0x1, 0xffffffff, 0xffffffff, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0};
+
+#endif  // 64BIT
+
+
+#if defined(P384_USE_S2N_BIGNUM_FIELD_ARITH)
 
 #if defined(OPENSSL_X86_64)
 // On x86_64 platforms s2n-bignum uses bmi2 and adx instruction sets
-// for some of the functions so we have to check if they are available.
-static inline uint8_t use_s2n_bignum(void) {
-  return ((OPENSSL_ia32cap_P[2] & (1u <<  8)) != 0) && // bmi2
-         ((OPENSSL_ia32cap_P[2] & (1u << 19)) != 0);   // adx
+// for some of the functions. These instructions are not supported by
+// every x86 CPU so we have to check if they are available and in case
+// they are not we fallback to slightly slower but generic implementation.
```

Comment:
Fiat? I dunno if it is even possible to switch between fiat and bignum at run-time anyway. It would both increase the binary size and complicate the code I suppose.

Reply:
Ai ai by generic I thought you meant portable C...